ygot/diff_test.go

```diff
@@ -0,0 +1,263 @@
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ygot
+
+import (
+	"fmt"
+	"reflect"
+	"testing"
+
+	"github.com/golang/protobuf/proto"
+	"github.com/kylelemons/godebug/pretty"
+	gnmipb "github.com/openconfig/gnmi/proto/gnmi"
+	"github.com/openconfig/ygot/util"
+)
+
+func TestSchemaPathToGNMIPath(t *testing.T) {
+	tests := []struct {
+		desc string
+		in   []string
+		want *gnmipb.Path
+	}{{
+		desc: "single element",
+		in:   []string{"one"},
+		want: &gnmipb.Path{
+			Elem: []*gnmipb.PathElem{{
+				Name: "one",
+			}},
+		},
+	}, {
+		desc: "multiple elements",
+		in:   []string{"one", "two", "three"},
+		want: &gnmipb.Path{
+			Elem: []*gnmipb.PathElem{{
+				Name: "one",
+			}, {
+				Name: "two",
+			}, {
+				Name: "three",
+			}},
+		},
+	}}
+
+	for _, tt := range tests {
+		if got := schemaPathTogNMIPath(tt.in); !proto.Equal(got, tt.want) {
+			t.Errorf("%s: schemaPathTogNMIPath(%v): did not get expected path, got: %v, want: %v", tt.desc, tt.in, pretty.Sprint(got), pretty.Sprint(tt.want))
+		}
+	}
+}
+
+func TestJoingNMIPaths(t *testing.T) {
+	tests := []struct {
+		desc     string
+		inParent *gnmipb.Path
+		inChild  *gnmipb.Path
+		want     *gnmipb.Path
+	}{{
+		desc: "simple parent and child",
+		inParent: &gnmipb.Path{
+			Elem: []*gnmipb.PathElem{{
+				Name: "one",
+			}},
+		},
+		inChild: &gnmipb.Path{
+			Elem: []*gnmipb.PathElem{{
+				Name: "two",
+			}},
+		},
+		want: &gnmipb.Path{
+			Elem: []*gnmipb.PathElem{{
+				Name: "one",
+			}, {
+				Name: "two",
+			}},
+		},
+	}, {
+		desc: "simple parent with list in child",
+		inParent: &gnmipb.Path{
+			Elem: []*gnmipb.PathElem{{
+				Name: "one",
+			}},
+		},
+		inChild: &gnmipb.Path{
+			Elem: []*gnmipb.PathElem{{
+				Name: "two",
+			}, {
+				Name: "three",
+				Key:  map[string]string{"four": "five"},
+			}},
+		},
+		want: &gnmipb.Path{
+			Elem: []*gnmipb.PathElem{{
+				Name: "one",
+			}, {
+				Name: "two",
+			}, {
+				Name: "three",
+				Key:  map[string]string{"four": "five"},
+			}},
+		},
+	}, {
+		desc: "list in parent, simple child",
+		inParent: &gnmipb.Path{
+			Elem: []*gnmipb.PathElem{{
+				Name: "one",
+				Key:  map[string]string{"two": "three"},
+			}},
+		},
+		inChild: &gnmipb.Path{
+			Elem: []*gnmipb.PathElem{{
+				Name: "four",
+			}},
+		},
+		want: &gnmipb.Path{
+			Elem: []*gnmipb.PathElem{{
+				Name: "one",
+				Key:  map[string]string{"two": "three"},
+			}, {
+				Name: "four",
+			}},
+		},
+	}}
+
+	for _, tt := range tests {
+		if got := joingNMIPaths(tt.inParent, tt.inChild); !proto.Equal(got, tt.want) {
+			diff := pretty.Compare(got, tt.want)
```

Comment:
(optional) Inside Google we usually pass want as the first argument to comparators like pretty to give the opposite direction (-want +got). Feel free to ignore, its just something I got used to.

Reply:
Yeah - unfortunately we're the other way around in this code. The "got before want" rule and the approach in `pretty.Compare` are the same throughout. At some point I can go back and fix this.